wordpress-mobile/WordPress-iOS
# Reader Comments: Implement notification settings sheet

Refs `pctCYC-9x-p2`
Depends on #17339

This PR adds the bottom sheet UI, along with the functionality. With this, the Follow Conversation via Notifications should be fully functional. Here are a couple of notes:

- It was a bit tricky to get the view controller to work with the `BottomSheetViewController`, as the height calculation doesn't seem to match properly. To solve this, I had to do a manual calculation – but thankfully nothing is hardcoded. Let me know if you think there's a better approach! 🙂 
- There's a bit of back-and-forth between the ObjC and the Swift version of `ReaderCommentsViewController` since the `BottomSheetViewController` could only be used from Swift (without changing the bottom sheet component).
- The feature flag is still disabled. It will be enabled in the next PR, along with some Tracks to track data on notification enable/disable actions.

Designs:

<img width="627" alt="Screen Shot 2021-10-20 at 04 45 05" src="https://user-images.githubusercontent.com/1299411/137995160-0f00cdbf-cebd-433f-aedc-e041e98b7767.png">

## To Test

Make sure that the `Follow Conversation via Notifications` flag is enabled.

The best way to test this is by using the [WordPress developer console](https://developer.wordpress.com/docs/api/console/), or some kind of API client (e.g. Postman) side by side with the app. The idea is to make changes on the iOS app, and then verify the values returned from `GET v1.1/sites/$site/posts/$post_id/subscribers/mine`:
```json
{
  "i_subscribe": false,
  "receives_notifications": false,
}
```

- Go to Reader, and select an unsubscribed post.
- Tap on the Follow button.
- Verify that the API returns `{ "i_subscribe": true, "receives_notifications": false }`.
- Tap on the "Enable" button. 
- Verify that the API returns `{ "i_subscribe": true, "receives_notifications": true }`.
- Tap the "Undo" button.
- Verify that the API now returns `{ "i_subscribe": true, "receives_notifications": false }`.
- Tap on the bell icon.
- Tap on the notification switch button, to enable notifications.
- Verify that the API returns `{ "i_subscribe": true, "receives_notifications": true }`.
- Now, turn the switch off.
- Verify that the API returns `{ "i_subscribe": true, "receives_notifications": false }`.
- Tap on the "Unfollow conversation" button.
- Verify that the API now returns `{ "i_subscribe": false, "receives_notifications": false }`.
- Verify that the bell icon button now turns into "Follow".

## Regression Notes
1. Potential unintended areas of impact
n/a. The feature is still hidden behind a flag.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
n/a. The feature is still hidden behind a flag.

3. What automated tests I added (or what prevented me from doing so)
n/a. The feature is still hidden behind a flag.

PR submission checklist:
